src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Debugging/RazorProximityExpressionResolver.cs

```diff
@@ -20,19 +24,78 @@ namespace Microsoft.VisualStudio.Razor.LanguageClient.Debugging;
+    IRemoteServiceInvoker remoteServiceInvoker) : IRazorProximityExpressionResolver
 {
-    private record CacheKey(Uri DocumentUri, long? HostDocumentSyncVersion, int Line, int Character);
+    private record CohostCacheKey(DocumentId DocumentId, VersionStamp Version, int Line, int Character) : CacheKey;
+    private record LspCacheKey(Uri DocumentUri, long? HostDocumentSyncVersion, int Line, int Character) : CacheKey;
+    private record CacheKey;
 
     private readonly FileUriProvider _fileUriProvider = fileUriProvider;
     private readonly LSPDocumentManager _documentManager = documentManager;
     private readonly ILSPProximityExpressionsProvider _proximityExpressionsProvider = proximityExpressionsProvider;
+    private readonly LanguageServerFeatureOptions _languageServerFeatureOptions = languageServerFeatureOptions;
+    private readonly IRemoteServiceInvoker _remoteServiceInvoker = remoteServiceInvoker;
 
     // 10 is a magic number where this effectively represents our ability to cache the last 10 "hit" breakpoint locations
     // corresponding proximity expressions which enables us not to go "async" in those re-hit scenarios.
     private readonly MemoryCache<CacheKey, IReadOnlyList<string>> _cache = new(sizeLimit: 10);
 
-    public async Task<IReadOnlyList<string>?> TryResolveProximityExpressionsAsync(ITextBuffer textBuffer, int lineIndex, int characterIndex, CancellationToken cancellationToken)
+    public Task<IReadOnlyList<string>?> TryResolveProximityExpressionsAsync(ITextBuffer textBuffer, int lineIndex, int characterIndex, CancellationToken cancellationToken)
+    {
+        if (_languageServerFeatureOptions.UseRazorCohostServer)
+        {
+            return TryResolveProximityExpressionsViaCohostingAsync(textBuffer, lineIndex, characterIndex, cancellationToken);
+        }
+
+        return TryResolveProximityExpressionsViaLspAsync(textBuffer, lineIndex, characterIndex, cancellationToken);
+
+    }
+
+    private async Task<IReadOnlyList<string>?> TryResolveProximityExpressionsViaCohostingAsync(ITextBuffer textBuffer, int lineIndex, int characterIndex, CancellationToken cancellationToken)
+    {
+        if (!textBuffer.TryGetTextDocument(out var razorDocument))
+        {
+            // Razor document is not in the Roslyn workspace.
+            return null;
+        }
+
+        if (razorDocument!.TryGetTextVersion(out var version))
+        {
+            version = await razorDocument.GetTextVersionAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        var cacheKey = new CohostCacheKey(razorDocument.Id, version, lineIndex, characterIndex);
+        if (_cache.TryGetValue(cacheKey, out var cachedRange))
+        {
+            // We've seen this request before. Hopefully the TryGetTextVersion call above was successful so this whole path
+            // will have been sync, and the cache will have been useful.
+            return cachedRange;
+        }
```

Comment:
Nit: Seems like mostly the same as first part of equivalent method of RazorBreakpointResolver? No huge savings, but I wonder if a generic helper method (with ChacheKey subtype being type parameter) or maybe even a simple base class would make sense. No strong feelings here, just seeing almost identical code as I'm reading it.

Reply:
I think if there were more of these sorts of things, I could see us having two different interfaces (ie `ILspBreakpointResolver` and `ICohostBreakpointResolver`) and each impl could share a base class etc. but I'm not sure it makes sense for just two. Perhaps in future if something like this comes up we could do something interesting with some custom MEF attributes so that the right impl gets magically imported.